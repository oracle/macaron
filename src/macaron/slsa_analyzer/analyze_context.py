# Copyright (c) 2022 - 2023, Oracle and/or its affiliates. All rights reserved.
# Licensed under the Universal Permissive License v 1.0 as shown at https://oss.oracle.com/licenses/upl/.

"""This module contains the Analyze Context class.

The AnalyzeContext is used to store the data of the repository being analyzed.
"""

import logging
import os
from typing import TypedDict

from macaron.database.table_definitions import Component, SLSALevel
from macaron.slsa_analyzer.checks.check_result import CheckResult, CheckResultType
from macaron.slsa_analyzer.git_service import BaseGitService
from macaron.slsa_analyzer.git_service.base_git_service import NoneGitService
from macaron.slsa_analyzer.levels import SLSALevels
from macaron.slsa_analyzer.provenance.expectations.expectation import Expectation
from macaron.slsa_analyzer.slsa_req import ReqName, SLSAReq, get_requirements_dict
from macaron.slsa_analyzer.specs.build_spec import BuildSpec
from macaron.slsa_analyzer.specs.ci_spec import CIInfo
from macaron.slsa_analyzer.specs.package_registry_spec import PackageRegistryInfo

logger: logging.Logger = logging.getLogger(__name__)


class ChecksOutputs(TypedDict):
    """Data computed at runtime by checks."""

    git_service: BaseGitService
    """The git service information for this repository."""
    build_spec: BuildSpec
    """The build spec inferred for this repository."""
    ci_services: list[CIInfo]
    """The CI services information for this repository."""
    is_inferred_prov: bool
    """True if we cannot find the provenance and Macaron need to infer the provenance."""
    # We need to use typing.Protocol for multiple inheritance, however, the Expectation
    # class uses inlined functions, which is not supported by Protocol.
    expectation: Expectation | None
    """The expectation to verify the provenance for this repository."""
    package_registries: list[PackageRegistryInfo]
    """The package registries for this repository."""


class AnalyzeContext:
    """This class contains data of the current analyzed repository."""

    def __init__(
        self,
        component: Component,
        macaron_path: str = "",
        output_dir: str = "",
    ):
        """Initialize instance.

        Parameters
        ----------
        component: Component
            The target software component.
        macaron_path : str
            The Macaron's root path.
        output_dir : str
            The output dir.
        """
        self.component = component
        self.ctx_data: dict[ReqName, SLSAReq] = get_requirements_dict()

        self.slsa_level = SLSALevels.LEVEL0
        # Indicate whether this repo fully reach a level or
        # it's just compliant for a SLSA level
        self.is_full_reach = False

        # The Macaron root path where the wrapper files exist.
        self.macaron_path = macaron_path

        # The output dir to store all files
        self.output_dir = output_dir

        # The check results from the analysis
        self.check_results: dict[str, CheckResult] = {}

        # Add the data computed at runtime to the dynamic_data attribute.
        self.dynamic_data: ChecksOutputs = ChecksOutputs(
            git_service=NoneGitService(),
            build_spec=BuildSpec(tools=[]),
            ci_services=[],
            package_registries=[],
            is_inferred_prov=True,
            expectation=None,
        )

    @property
    def provenances(self) -> dict:
        """Return the provenances data as a dictionary.

        Returns
        -------
        dict
            A dictionary in which each key is a CI service's name and each value is
            the corresponding provenance payload.
        """
        try:
            ci_services = self.dynamic_data["ci_services"]
            result = {}
            for ci_info in ci_services:
                result[ci_info["service"].name] = [payload.statement for payload in ci_info["provenances"]]
            package_registry_entries = self.dynamic_data["package_registries"]
            for package_registry_entry in package_registry_entries:
                result[package_registry_entry.package_registry.name] = [
                    provenance.payload.statement for provenance in package_registry_entry.provenances
                ]
            return result
        except KeyError:
            return {}

    # TODO: refactor as this information is related to the reporter not analyze context
    @property
    def is_inferred_provenance(self) -> bool:
        """Return True if the provenance for this repo is an inferred one.

        Returns
        -------
        bool
        """
        return self.dynamic_data["is_inferred_prov"]

    def update_req_status(self, req_name: ReqName, status: bool, feedback: str) -> None:
        """Update the status of a single requirement.

        Parameters
        ----------
        req_name : ReqName
            The requirement to update.
        status : bool
            True if the requirement passes, else False.
        feedback: str
            The feedback to the requirement.
        """
        req = self.ctx_data.get(req_name)
        if req:
            logger.debug(
                "Update requirement %s: set to %s (%s)",
                req_name.value,
                status,
                feedback,
            )
            self.ctx_data[req_name].set_status(status, feedback)
        else:
            logger.debug("Trying to update non-existing requirement, ignoring ...")

    def bulk_update_req_status(self, req_list: list, status: bool, feedback: str) -> None:
        """Update the status of a requirements in ``req_list``.

        Parameters
        ----------
        req_list : list[ReqName]
            The list of requirement to update.
        status : bool
            True if the requirement passes, else False.
        feedback : str
            The feedback to the requirement.
        """
        for req in req_list:
            self.update_req_status(req, status, feedback)

    def get_slsa_level_table(self) -> SLSALevel:
        """Return filled ORM table storing the level for this component."""
        # TODO: right now `slsa_level` is always 0 and `is_full_reach` is False,
        # which needs to be handled properly.
        return SLSALevel(
            component=self.component,
            slsa_level=int(self.slsa_level),
            reached=self.is_full_reach,
        )

    def get_dict(self) -> dict:
        """Return the dictionary representation of the AnalyzeContext instance."""
        _sorted_on_id = sorted(self.check_results.values(), key=lambda item: item["check_id"])
        # Remove result_tables since we don't have a good json representation for them.
        sorted_on_id = []
        for res in _sorted_on_id:
            # res is CheckResult(TypedDict)
            res: dict = dict(res.copy())  # type: ignore
            res.pop("result_tables")  # type: ignore
            sorted_on_id.append(res)
        sorted_results = sorted(sorted_on_id, key=lambda item: item["result_type"], reverse=True)
        check_summary = {
            result_type.value: len(result_list) for result_type, result_list in self.get_check_summary().items()
        }
        check_summary_sorted = dict(sorted(check_summary.items()))
        result = {
            "info": {
                "full_name": self.component.purl,
                "local_cloned_path": os.path.relpath(self.component.repository.fs_path, self.output_dir)
                if self.component.repository
                else "Unable to find a repository.",
                "remote_path": self.component.repository.remote_path if self.component.repository else "",
                "branch": self.component.repository.branch_name if self.component.repository else "",
                "commit_hash": self.component.repository.commit_sha if self.component.repository else "",
                "commit_date": self.component.repository.commit_date if self.component.repository else "",
            },
            "provenances": {
                "is_inferred": self.is_inferred_provenance,
                "content": self.provenances,
            },
            "checks": {"summary": check_summary_sorted, "results": sorted_results},
        }
        return result

    def get_check_summary(self) -> dict[CheckResultType, list[CheckResult]]:
        """Return the summary of all checks results for the target repository.

        Returns
        -------
        dict[CheckResultType, list[CheckResult]]
            The mapping of the check result type and the related check results.
        """
        result: dict[CheckResultType, list[CheckResult]] = {result_type: [] for result_type in CheckResultType}

        for check_result in self.check_results.values():
            match check_result["result_type"]:
                case CheckResultType.PASSED:
                    result[CheckResultType.PASSED].append(check_result)
                case CheckResultType.SKIPPED:
                    result[CheckResultType.SKIPPED].append(check_result)
                case CheckResultType.FAILED:
                    result[CheckResultType.FAILED].append(check_result)
                case CheckResultType.DISABLED:
                    result[CheckResultType.DISABLED].append(check_result)
                case CheckResultType.UNKNOWN:
                    result[CheckResultType.UNKNOWN].append(check_result)

        return result

    def __str__(self) -> str:
        """Return the string representation of the AnalyzeContext instance."""
        output = ""

        check_summary = self.get_check_summary()

        for check_result in self.check_results.values():
            output = "".join(
                [
                    output,
                    f"Check {check_result['check_id']}: {check_result['check_description']}\n",
                    f"\t{check_result['result_type'].value}\n",
                ]
            )

        for result_type, result_list in check_summary.items():
            output = "".join([output, f"{len(result_list)} checks {result_type.value}\n"])

        return output
