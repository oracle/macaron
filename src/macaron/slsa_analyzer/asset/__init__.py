# Copyright (c) 2023 - 2023, Oracle and/or its affiliates. All rights reserved.
# Licensed under the Universal Permissive License v 1.0 as shown at https://oss.oracle.com/licenses/upl/.

"""This module defines classes and interfaces related to assets.

Assets are files published from some build.
"""

from typing import Protocol


class AssetLocator(Protocol):
    """Interface of an asset locator."""

    @property
    def name(self) -> str:
        """Get the name (file name) of the asset."""

    @property
    def url(self) -> str:
        """Get the url to the asset."""

    @property
    def size_in_bytes(self) -> int:
        """Get the size of the asset in bytes."""

    def download(self, dest: str) -> bool:
        """Download the asset.

        Parameters
        ----------
        dest : str
            The local destination where the asset is downloaded to.
            Note that this must include the file name.

        Returns
        -------
        bool
            ``True`` if the asset is downloaded successfully; ``False`` if not.
        """
