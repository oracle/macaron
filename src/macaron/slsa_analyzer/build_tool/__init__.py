# Copyright (c) 2022 - 2022, Oracle and/or its affiliates. All rights reserved.
# Licensed under the Universal Permissive License v 1.0 as shown at https://oss.oracle.com/licenses/upl/.

"""The build_tool package contains the supported build tools for Macaron."""

from .base_build_tool import BaseBuildTool
from .gradle import Gradle
from .maven import Maven

# The list of supported build tools. The order of the list determine the order
# in which each build tool is checked against the target repository.
BUILD_TOOLS: list[BaseBuildTool] = [Gradle(), Maven()]
