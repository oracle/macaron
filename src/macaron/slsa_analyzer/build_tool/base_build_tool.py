# Copyright (c) 2022 - 2024, Oracle and/or its affiliates. All rights reserved.
# Licensed under the Universal Permissive License v 1.0 as shown at https://oss.oracle.com/licenses/upl/.

"""This module contains the BaseBuildTool class to be inherited by other specific Build Tools."""

import glob
import itertools
import json
import logging
import os
from abc import ABC, abstractmethod
from collections.abc import Iterable
from dataclasses import dataclass
from pathlib import Path
from typing import TypedDict

from macaron.code_analyzer.call_graph import BaseNode
from macaron.dependency_analyzer import DependencyAnalyzer
from macaron.slsa_analyzer.build_tool.language import BuildLanguage
from macaron.slsa_analyzer.checks.check_result import Confidence, Evidence, EvidenceWeightMap

logger: logging.Logger = logging.getLogger(__name__)


class BuildToolCommand(TypedDict):
    """This class is an abstraction for build tool commands storing useful contextual data for analysis."""

    #: The parsed build tool command. This command can be any bash command whose program name is the build tool.
    command: list[str]

    #: The name of the language to build the artifact.
    language: str

    #: The list of possible language version numbers.
    language_versions: list[str] | None

    #: The list of possible language distributions.
    language_distributions: list[str] | None

    #: The URL providing information about the language distributions and versions.
    language_url: str | None

    #: The relative path to the root CI file that ultimately triggers the command.
    ci_path: str

    #: The CI step object that calls the command.
    step_node: BaseNode

    #: The list of name of reachable variables that contain secrets."""
    reachable_secrets: list[str]

    #: The name of CI events that trigger the workflow running the build command.
    events: list[str] | None


def file_exists(path: str, file_name: str) -> bool:
    """Return True if a file exists in a directory.

    This method searches in the directory recursively.

    Parameters
    ----------
    path : str
        The path to search for the file.
    file_name : str
        The name of the file to search.

    Returns
    -------
    bool
        True if file_name exists else False.
    """
    pattern = os.path.join(path, "**", file_name)
    files_detected = glob.iglob(pattern, recursive=True)
    try:
        next(files_detected)
        return True
    except StopIteration:
        return False


@dataclass
class RuntimeOptions:
    """The class for build tool runtime configurations read from `defaults.ini`.

    Note that Macaron uses the options in this class to "run" a build tool.
    """

    #: The timeout used for running the build tool commands.
    build_timeout: float = 600


class BaseBuildTool(ABC):
    """This abstract class is used to implement Build Tools."""

    def __init__(self, name: str, language: BuildLanguage) -> None:
        """Initialize instance.

        Parameters
        ----------
        name : str
            The name of this build tool.
        language: BuildLanguage
            The name of the language used by the programs built by the build tool.
        """
        self.name = name
        self.language = language
        self.entry_conf: list[str] = []
        self.build_configs: list[str] = []
        self.package_lock: list[str] = []
        self.builder: list[str] = []
        self.packager: list[str] = []
        self.publisher: list[str] = []
        self.interpreter: list[str] = []
        self.interpreter_flag: list[str] = []
        self.build_arg: list[str] = []
        self.deploy_arg: list[str] = []
        self.ci_build_kws: dict[str, list[str]] = {
            "github_actions": [],
            "travis_ci": [],
            "circle_ci": [],
            "gitlab_ci": [],
            "jenkins": [],
        }
        self.ci_deploy_kws: dict[str, list[str]] = {
            "github_actions": [],
            "travis_ci": [],
            "circle_ci": [],
            "gitlab_ci": [],
            "jenkins": [],
        }
        self.build_log: list[str] = []
        self.wrapper_files: list[str] = []
        self.runtime_options = RuntimeOptions()

    def __str__(self) -> str:
        return self.name

    @abstractmethod
    def is_detected(self, repo_path: str) -> bool:
        """Return True if this build tool is used in the target repo.

        Parameters
        ----------
        repo_path : str
            The path to the target repo.

        Returns
        -------
        bool
            True if this build tool is detected, else False.
        """

    @abstractmethod
    def prepare_config_files(self, wrapper_path: str, build_dir: str) -> bool:
        """Prepare the necessary wrapper files for running the build.

        This method will return False if there is any errors happened during operation.

        Parameters
        ----------
        wrapper_path : str
            The path where all necessary wrapper files are located.
        build_dir : str
            The path of the build dir. This is where all files are copied to.

        Returns
        -------
        bool
            True if succeed else False.
        """

    @abstractmethod
    def load_defaults(self) -> None:
        """Load the default values from defaults.ini."""

    @abstractmethod
    def get_dep_analyzer(self, repo_path: str) -> DependencyAnalyzer:
        """Create a DependencyAnalyzer for the build tool.

        Parameters
        ----------
        repo_path: str
            The path to the target repo.

        Returns
        -------
        DependencyAnalyzer
            The DependencyAnalyzer object.
        """

    def get_build_dirs(self, repo_path: str) -> Iterable[Path]:
        """Find directories in the repository that have their own build scripts.

        This is especially important for applications that consist of multiple services.

        Parameters
        ----------
        repo_path: str
            The path to the target repo.

        Yields
        ------
        Path
            The relative paths from the repo path that contain build scripts.
        """
        config_paths: set[str] = set()
        for build_cfg in self.build_configs:
            config_paths.update(
                path
                for path in glob.glob(os.path.join(repo_path, "**", build_cfg), recursive=True)
                if self.is_detected(str(Path(path).parent))
            )

        list_iter = iter(sorted(config_paths, key=lambda x: (str(Path(x).parent), len(Path(x).parts))))
        try:
            cfg_path = next(list_iter)
            yield Path(cfg_path).parent.relative_to(repo_path)
            while next_item := next(list_iter):
                if next_item.startswith(str(Path(cfg_path).parent)):
                    continue
                cfg_path = next_item
                yield Path(next_item).parent.relative_to(repo_path)

        except StopIteration:
            pass

    def serialize_to_json(self, cmd: list[str]) -> str:
        """Convert a list of values to a json-encoded string so that it is easily parsable by later consumers.

        Parameters
        ----------
        cmd: list[str]
            List of command-line arguments.

        Returns
        -------
        str
            The list of command-line arguments as a json-encoded string.
        """
        return json.dumps(cmd)

    def is_build_command(self, cmd: list[str]) -> bool:
        """
        Determine if the command is a build tool command.

        Parameters
        ----------
        cmd: list[str]
            List of command-line arguments.

        Returns
        -------
        bool
            True if the command is a build tool command.
        """
        # Check for empty or invalid commands.
        if not cmd or not cmd[0]:
            return False
        # The first argument in a bash command is the program name.
        # So first check that the program name is a supported build tool name.
        # We need to handle cases where the first argument is a path to the program.
        cmd_program_name = os.path.basename(cmd[0])
        if not cmd_program_name:
            logger.debug("Found invalid program name %s.", cmd[0])
            return False

        build_tools = set(itertools.chain(self.builder, self.packager, self.publisher, self.interpreter))
        if any(tool for tool in build_tools if tool == cmd_program_name):
            return True

        return False

    def match_cmd_args(self, cmd: list[str], tools: list[str], args: list[str]) -> bool:
        """
        Check if the build command matches any of the tools and the command-line arguments.

        If build command's first element, which is the program name matches any of the `tools` names and any of its arguments
        match any of the arguments in `args`, this function returns True.

        Parameters
        ----------
        cmd: list[str]
            The command-line arguments.
        tools: list[str]
            The name of tools that will be matched with the program name in the bash command.
        args: list[str]
            The lit of arguments that should match with the bash command.

        Returns
        -------
        bool
            True if the provided command matches the tool and arguments.
        """
        cmd_program_name = os.path.basename(cmd[0])

        if cmd_program_name in tools:
            # Check the arguments in the bash command.
            # If there are no args expected for this build tool, accept the command.
            if not args:
                logger.debug("No build arguments required. Accept the %s command.", self.serialize_to_json(cmd))
                return True

            for word in cmd[1:]:
                # TODO: allow plugin versions in arguments, e.g., maven-plugin:1.6.8:deploy.
                if word in args:
                    logger.debug("Found the command %s.", self.serialize_to_json(cmd))
                    return True

        return False

    def infer_confidence_deploy_command(self, cmd: BuildToolCommand) -> Confidence:
        """
        Infer the confidence level for the deploy command.

        Parameters
        ----------
        cmd: BuildToolCommand
            The build tool command object.

        Returns
        -------
        Confidence
            The confidence level for the deploy command.
        """
        # Apply heuristics and assign weights and scores for the discovered evidence.
        # TODO: infer the scores based on existing data using probabilistic inference.
        # Initialize the map.
        evidence_weight_map = EvidenceWeightMap(
            [
                Evidence(name="reachable_secrets", found=False, weight=1),
                Evidence(name="ci_workflow_name", found=False, weight=2),
                Evidence(name="release_event", found=False, weight=2),
            ]
        )

        # Check if secrets are present in the caller job.
        if cmd["reachable_secrets"]:
            evidence_weight_map.update_result(name="reachable_secrets", found=True)

        # Check workflow names.
        deploy_keywords = ["release", "deploy", "publish"]
        for kw in deploy_keywords:
            if kw in os.path.basename(cmd["ci_path"]).lower():
                evidence_weight_map.update_result(name="ci_workflow_name", found=True)
                break

        # Check if the CI workflow is triggered by a release event.
        if cmd["events"] and "release" in cmd["events"]:
            evidence_weight_map.update_result(name="release_event", found=True)

        return Confidence.normalize(evidence_weight_map=evidence_weight_map)

    def is_deploy_command(
        self, cmd: BuildToolCommand, excluded_configs: list[str] | None = None
    ) -> tuple[bool, Confidence]:
        """
        Determine if the command is a deploy command.

        A deploy command usually performs multiple tasks, such as compilation, packaging, and publishing the artifact.
        This function filters the build tool commands that are called from the configuration files provided as input.

        Parameters
        ----------
        cmd: BuildToolCommand
            The build tool command object.
        excluded_configs: list[str] | None
            Build tool commands that are called from these configuration files are excluded.

        Returns
        -------
        tuple[bool, Confidence]
            Return True along with the inferred confidence level if the command is a deploy tool command.
        """
        # Check the language.
        if cmd["language"] is not self.language:
            return False, Confidence.HIGH
        # Some projects use a publisher tool and some use the build tool with deploy arguments.
        deploy_tool = self.publisher if self.publisher else self.builder

        if not self.match_cmd_args(cmd=cmd["command"], tools=deploy_tool, args=self.deploy_arg):
            return False, Confidence.HIGH

        # Check if the CI workflow is a configuration for a known tool.
        if excluded_configs and os.path.basename(cmd["ci_path"]) in excluded_configs:
            return False, Confidence.HIGH

        return True, self.infer_confidence_deploy_command(cmd=cmd)

    def is_package_command(
        self, cmd: BuildToolCommand, excluded_configs: list[str] | None = None
    ) -> tuple[bool, Confidence]:
        """
        Determine if the command is a packaging command.

        A packaging command usually performs multiple tasks, such as compilation and creating the artifact.
        This function filters the build tool commands that are called from the configuration files provided as input.

        Parameters
        ----------
        cmd: BuildToolCommand
            The build tool command object.
        excluded_configs: list[str] | None
            Build tool commands that are called from these configuration files are excluded.

        Returns
        -------
        tuple[bool, Confidence]
            Return True along with the inferred confidence level if the command is a build tool command.
        """
        # Check the language.
        if cmd["language"] is not self.language:
            return False, Confidence.HIGH

        builder = self.packager if self.packager else self.builder

        if not self.match_cmd_args(cmd=cmd["command"], tools=builder, args=self.build_arg):
            return False, Confidence.HIGH

        # Check if the CI workflow is a configuration for a known tool.
        if excluded_configs and os.path.basename(cmd["ci_path"]) in excluded_configs:
            return False, Confidence.HIGH

        return True, Confidence.HIGH
