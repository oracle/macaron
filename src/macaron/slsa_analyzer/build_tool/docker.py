# Copyright (c) 2023 - 2024, Oracle and/or its affiliates. All rights reserved.
# Licensed under the Universal Permissive License v 1.0 as shown at https://oss.oracle.com/licenses/upl/.

"""This module contains the Docker class which inherits BaseBuildTool.

This module is used to work with repositories that use Docker as a build tool.
"""

from macaron.config.defaults import defaults
from macaron.dependency_analyzer.dependency_resolver import NoneDependencyAnalyzer
from macaron.slsa_analyzer.build_tool.base_build_tool import BaseBuildTool, file_exists
from macaron.slsa_analyzer.build_tool.language import BuildLanguage


class Docker(BaseBuildTool):
    """This class contains the information of Docker when used as a build tool."""

    def __init__(self) -> None:
        """Initialize instance."""
        super().__init__(name="docker", language=BuildLanguage.DOCKER)

    def load_defaults(self) -> None:
        """Load the default values from defaults.ini."""
        if "builder.docker" in defaults:
            for item in defaults["builder.docker"]:
                if hasattr(self, item):
                    setattr(self, item, defaults.get_list("builder.docker", item))

        if "builder.docker.ci.deploy" in defaults:
            for item in defaults["builder.docker.ci.deploy"]:
                if item in self.ci_deploy_kws:
                    self.ci_deploy_kws[item] = defaults.get_list("builder.docker.ci.deploy", item)

    def is_detected(self, repo_path: str) -> bool:
        """Return True if this build tool is used in the target repo.

        Parameters
        ----------
        repo_path : str
            The path to the target repo.

        Returns
        -------
        bool
            True if this build tool is detected, else False.
        """
        for file in self.build_configs:
            if file_exists(repo_path, file):
                return True

        return False

    def prepare_config_files(self, wrapper_path: str, build_dir: str) -> bool:
        """Make necessary preparations for using this build tool.

        Parameters
        ----------
        wrapper_path : str
            The path where all necessary wrapper files are located.
        build_dir : str
            The path of the build dir. This is where all files are copied to.

        Returns
        -------
        bool
            True if succeed else False.
        """
        # TODO: Future dependency analysis may require some preprocessing, e.g.
        #       saving images to tar files. Need to investigate when implementing
        #       and work with this method accordingly.

        return False

    def get_dep_analyzer(self, repo_path: str) -> NoneDependencyAnalyzer:
        """Create a DependencyAnalyzer for the Docker build tool. Currently unimplemented.

        Parameters
        ----------
        repo_path: str
            The path to the target repo.

        Returns
        -------
        NoneDependencyAnalyser
            The NoneDependencyAnalyser object.

        Raises
        ------
        DependencyAnalyzerError
        """
        # TODO: Find a suitable tool to analyse dependencies; as of now Syft
        #       seems to be a good option, but need to experiment.
        return NoneDependencyAnalyzer()
