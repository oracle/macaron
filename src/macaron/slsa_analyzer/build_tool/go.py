# Copyright (c) 2023 - 2024, Oracle and/or its affiliates. All rights reserved.
# Licensed under the Universal Permissive License v 1.0 as shown at https://oss.oracle.com/licenses/upl/.

"""This module contains the Go class which inherits BaseBuildTool.

This module is used to work with repositories that have Go.
"""

from macaron.config.defaults import defaults
from macaron.dependency_analyzer.dependency_resolver import DependencyAnalyzer, NoneDependencyAnalyzer
from macaron.slsa_analyzer.build_tool.base_build_tool import BaseBuildTool, file_exists
from macaron.slsa_analyzer.build_tool.language import BuildLanguage


class Go(BaseBuildTool):
    """This class contains the information of the Go build tool."""

    def __init__(self) -> None:
        super().__init__(name="go", language=BuildLanguage.GO)

    def load_defaults(self) -> None:
        """Load the default values from defaults.ini."""
        if "builder.go" in defaults:
            for item in defaults["builder.go"]:
                if hasattr(self, item):
                    setattr(self, item, defaults.get_list("builder.go", item))

        if "builder.go.ci.deploy" in defaults:
            for item in defaults["builder.go.ci.deploy"]:
                if item in self.ci_deploy_kws:
                    self.ci_deploy_kws[item] = defaults.get_list("builder.go.ci.deploy", item)

    def is_detected(self, repo_path: str) -> bool:
        """Return True if this build tool is used in the target repo.

        Parameters
        ----------
        repo_path : str
            The path to the target repo.

        Returns
        -------
        bool
            True if this build tool is detected, else False.
        """
        go_config_files = self.build_configs + self.entry_conf
        for file in go_config_files:
            if file_exists(repo_path, file):
                return True

        return False

    def prepare_config_files(self, wrapper_path: str, build_dir: str) -> bool:
        """Prepare the necessary wrapper files for running the build.

        Go doesn't require preparation, so return true.

        Parameters
        ----------
        wrapper_path : str
            The path where all necessary wrapper files are located.
        build_dir : str
            The path of the build dir. This is where all files are copied to.

        Returns
        -------
        bool
            True if succeed else False.
        """
        return True

    def get_dep_analyzer(self, repo_path: str) -> DependencyAnalyzer:
        """Create a DependencyAnalyzer for the build tool.

        Parameters
        ----------
        repo_path: str
            The path to the target repo.

        Returns
        -------
        DependencyAnalyzer
            The DependencyAnalyzer object.
        """
        # TODO: Implement this method.
        return NoneDependencyAnalyzer()
