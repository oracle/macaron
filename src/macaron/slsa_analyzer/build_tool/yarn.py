# Copyright (c) 2023 - 2023, Oracle and/or its affiliates. All rights reserved.
# Licensed under the Universal Permissive License v 1.0 as shown at https://oss.oracle.com/licenses/upl/.

"""This module contains the Yarn class which inherits BaseBuildTool.

This module is used to work with repositories that use Yarn as its
build tool.
"""

from macaron.config.defaults import defaults
from macaron.dependency_analyzer.dependency_resolver import DependencyAnalyzer, NoneDependencyAnalyzer
from macaron.slsa_analyzer.build_tool.base_build_tool import BaseBuildTool, file_exists


class Yarn(BaseBuildTool):
    """This class contains the information of the yarn build tool."""

    def __init__(self) -> None:
        super().__init__(name="yarn")

    def load_defaults(self) -> None:
        """Load the default values from defaults.ini."""
        if "builder.yarn" in defaults:
            for item in defaults["builder.yarn"]:
                if hasattr(self, item):
                    setattr(self, item, defaults.get_list("builder.yarn", item))

        # TODO: Find a suitable github action for Yarn
        # if "builder.yarn.ci.deploy" in defaults:
        #     for item in defaults["builder.yarn.ci.deploy"]:
        #         if item in self.ci_deploy_kws:
        #             self.ci_deploy_kws[item] = defaults.get_list("builder.yarn.ci.deploy", item)

    def is_detected(self, repo_path: str) -> bool:
        """Return True if this build tool is used in the target repo.

        Parameters
        ----------
        repo_path : str
            The path to the target repo.

        Returns
        -------
        bool
            True if this build tool is detected, else False.
        """
        # TODO: When more complex build detection is being implemented, consider
        #       cases like .yarnrc existing but not package-lock.json and whether
        #       they would still count as "detected"
        yarn_config_files = self.build_configs + self.package_lock + self.entry_conf
        for file in yarn_config_files:
            if file_exists(repo_path, file):
                return True

        return False

    def prepare_config_files(self, wrapper_path: str, build_dir: str) -> bool:
        """Prepare the necessary wrapper files for running the build.

        yarn doesn't require preparation, so return true.

        Parameters
        ----------
        wrapper_path : str
            The path where all necessary wrapper files are located.
        build_dir : str
            The path of the build dir. This is where all files are copied to.

        Returns
        -------
        bool
            True if succeed else False.
        """
        return True

    def get_dep_analyzer(self, repo_path: str) -> DependencyAnalyzer:
        """Create a DependencyAnalyzer for the build tool.

        Parameters
        ----------
        repo_path: str
            The path to the target repo.

        Returns
        -------
        DependencyAnalyzer
            The DependencyAnalyzer object.
        """
        # TODO: Implement this method.
        return NoneDependencyAnalyzer()
