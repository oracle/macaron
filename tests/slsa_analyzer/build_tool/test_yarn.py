# Copyright (c) 2023 - 2023, Oracle and/or its affiliates. All rights reserved.
# Licensed under the Universal Permissive License v 1.0 as shown at https://oss.oracle.com/licenses/upl/.

"""This module tests the Yarn build functions."""

from pathlib import Path

import pytest

from macaron.slsa_analyzer.build_tool.yarn import Yarn
from tests.slsa_analyzer.mock_git_utils import prepare_repo_for_testing


@pytest.mark.parametrize(
    "mock_repo",
    [
        Path(__file__).parent.joinpath("mock_repos", "yarn_repos", "root_package"),
        Path(__file__).parent.joinpath("mock_repos", "yarn_repos", "root_package_packagelock"),
        Path(__file__).parent.joinpath("mock_repos", "yarn_repos", "nested_package"),
        Path(__file__).parent.joinpath("mock_repos", "yarn_repos", "no_package"),
    ],
)
def test_get_build_dirs(snapshot: list, yarn_tool: Yarn, mock_repo: Path) -> None:
    """Test discovering build directories."""
    assert list(yarn_tool.get_build_dirs(str(mock_repo))) == snapshot


@pytest.mark.parametrize(
    ("mock_repo", "expected_value"),
    [
        (Path(__file__).parent.joinpath("mock_repos", "yarn_repos", "root_package"), True),
        (Path(__file__).parent.joinpath("mock_repos", "yarn_repos", "root_package_packagelock"), True),
        (Path(__file__).parent.joinpath("mock_repos", "yarn_repos", "nested_package"), True),
        (Path(__file__).parent.joinpath("mock_repos", "yarn_repos", "no_package"), False),
    ],
)
def test_yarn_build_tool(yarn_tool: Yarn, macaron_path: str, mock_repo: str, expected_value: bool) -> None:
    """Test the yarn build tool."""
    base_dir = Path(__file__).parent
    ctx = prepare_repo_for_testing(mock_repo, macaron_path, base_dir)
    assert yarn_tool.is_detected(ctx.component.repository.fs_path) == expected_value
