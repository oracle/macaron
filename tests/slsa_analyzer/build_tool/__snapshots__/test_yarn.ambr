# serializer version: 1
# name: test_get_build_dirs[mock_repo0]
  list([
    PosixPath('.'),
  ])
# ---
# name: test_get_build_dirs[mock_repo1]
  list([
    PosixPath('.'),
  ])
# ---
# name: test_get_build_dirs[mock_repo2]
  list([
    PosixPath('project'),
  ])
# ---
# name: test_get_build_dirs[mock_repo3]
  list([
  ])
# ---
